src/data/arena-tags/arena-tag.ts

```diff
@@ -0,0 +1,111 @@
+import { globalScene } from "#app/global-scene";
+import { allMoves } from "#data/data-lists";
+import { ArenaTagSide } from "#enums/arena-tag-side";
+import type { ArenaTagType } from "#enums/arena-tag-type";
+import type { MoveId } from "#enums/move-id";
+import type { Arena } from "#field/arena";
+import type { Pokemon } from "#field/pokemon";
+import i18next from "i18next";
+
+/**
+ * Base class for any special effects that apply to the {@linkcode Arena | field}
+ * during battle.
+ */
+export abstract class ArenaTag {
+  constructor(
+    /** An {@linkcode ArenaTagType | identifier} for the tag's effect. */
+    public tagType: ArenaTagType,
+    /**
+     * The number of turns the tag is active. If set to a number less than
+     * 1, the Arena Tag will last indefinitely (until its {@linkcode Arena} is reset).
+     */
+    public turnCount: number,
+    /** (Optional) The {@linkcode MoveId | move} that created this effect. */
+    public sourceMoveId?: MoveId,
+    /** (Optional) The ID number of the {@linkcode Pokemon} that created this effect. */
+    public sourceId?: number,
+    /**
+     * The {@linkcode ArenaTagSide} to which this effect applies. Tags can
+     * affect the player's and/or the enemy's side of the field.
+     */
+    public side: ArenaTagSide = ArenaTagSide.BOTH,
+  ) {}
+
+  /**
+   * Used to obtain localized keys for the tag's messages based on what
+   * side of the field the tag affects.
+   * @example
+   * ```ts
+   * i18next.t(`arenaTag:arenaOnRemove${this.i18nSideKey}`);
+   * ```
+   */
+  public get i18nSideKey(): string {
+    if (this.side === ArenaTagSide.PLAYER) {
+      return "Player";
+    }
+    if (this.side === ArenaTagSide.ENEMY) {
+      return "Enemy";
+    }
+    return "";
+  }
+
+  public apply(_arena: Arena, _simulated: boolean, ..._args: unknown[]): boolean {
+    return true;
+  }
+
+  public onAdd(_arena: Arena, _quiet: boolean = false): void {}
+
+  public onRemove(_arena: Arena, quiet: boolean = false): void {
+    if (!quiet) {
+      globalScene.phaseManager.queueMessagePhase(
+        i18next.t(`arenaTag:arenaOnRemove${this.i18nSideKey}`, { moveName: this.getMoveName() }),
+      );
+    }
+  }
+
+  public onOverlap(_arena: Arena): void {}
```

Comment:
Shouldn't this be `public abstract onOverlap(_arena: Arena): void` or is it optional to implement them?